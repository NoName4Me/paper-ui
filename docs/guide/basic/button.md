# Button



```js
import { PButton } from 'vue-paper-ui'

export default {
  components: {
    PButton
  }
}
```

<script>
  import { PButton } from '$paper-ui/index'
  export default {
    name: 'app',
    components: {
      PButton
    }
  }
</script>

<PButton>Button</PButton>
<PButton type="primary" size="large">Button</PButton>
<PButton size="small" @click="function() {alert('hi')}">Button</PButton>
<PButton type="primary" disabled>Button</PButton>
<PButton type="link">Button</PButton>

对应代码：

```html
<PButton>Button</PButton>
<PButton type="primary" size="large">Button</PButton>
<PButton size="small" @click="console.log('hi')">Button</PButton>
<PButton type="primary" disabled>Button</PButton>
<PButton type="link">Button</PButton>
```