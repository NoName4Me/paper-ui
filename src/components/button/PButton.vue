<template>
  <button class="p-btn"
          :class="[size, type]"
          :disabled="disabled"
          @click="handleClick">
    <i :class="icon"
       v-if="icon"></i>
    <slot></slot>
  </button>
</template>

<script lang="ts">
import { Vue, Component, Prop, Emit } from "vue-property-decorator";

@Component
export default class PButton extends Vue {
  @Prop({ default: "" }) private type!: string;
  @Prop({ default: "middle" }) private size!: string;
  @Prop({ default: "" }) private icon!: string;
  @Prop({ default: false }) private disabled!: boolean;

  // `methods`, emit an event
  public handleClick(event: MouseEvent) {
    if (this.disabled) {
      event.preventDefault();
      event.stopPropagation();
      return;
    }
    this.$emit("click", event);
  }
}
</script>

<style lang="scss">
@import "../../assets/keyframes.scss";
@import "../../assets/basics.scss";
// TODO: button active state style
@function scaleStyle($base, $factor) {
  @if length($base) == 1 {
    @return $base / $factor;
  }
  @if length($base) == 2 {
    @return nth($base, 1) / $factor nth($base, 2) / $factor;
  }
  // @return @each $val in $padding { $val / $factor }
}
.p-btn {
  $root: &;
  $baseFontSize: 12px;
  $padding: 8px 16px;
  
  padding: $padding;
  transition: all 0.2s ease;
  cursor: default;
  box-sizing: border-box;
  font-size: $baseFontSize;
  border: 1px solid #eee;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
  // reset default
  &:active,
  &:hover,
  &:focus {
    outline: none;
  }
  &:active {
    box-shadow: inset 0 1px 2px rgba(0, 0, 0, 0.2);
  }
  &.small {
    padding: scaleStyle($padding, 2);
  }
  &.large {
    padding: scaleStyle($padding, 0.8);
    font-size: scaleStyle($baseFontSize, 0.8);
  }
  &.primary {
    border: 1px solid $colorMain;
    //
  }
  &.info {
    border: 1px solid $colorInfo;
  }
  &.success {
    border: 1px solid $colorSuccess;
  }
  &.error {
    border: 1px solid $colorError;
  }
  &.warn {
    border: 1px solid $colorWarn;
  }

  &:hover:not(:disabled) {
    opacity: 0.9;
    box-shadow: 1px 4px 6px rgba(0, 0, 0, 0.2);
  }
  &:disabled {
    cursor: not-allowed;
    background-color: #efefef;
    box-shadow: inset 1px 1px 4px rgba(0, 0, 0, 0.2);
    &:hover {
      animation: shake 0.4s;
    }
  }
  &.link {
    border: none;
    text-decoration: underline;
    &:hover,
    &:active {
      border: none;
      box-shadow: none;
    }
    &:hover {
      font-weight: 500;
    }
  }
}
</style>

