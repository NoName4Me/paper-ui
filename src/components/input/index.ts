import PInput from './PInput.vue';
import Vue from 'vue';


export {
  PInput,
};

// Declare install function executed by Vue.use()
export default {
  install(vue: typeof Vue, options?: any) {
    if ((this as any).installed) {
      return;
    }
    
    (this as any).installed = true;
    vue.component('PInput', PInput);
  },
};
