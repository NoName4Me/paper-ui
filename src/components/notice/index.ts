import PNotice from './PNotice.vue';
import Vue from 'vue';

const Notice = Vue.extend(PNotice)
const cache = new Map<string, PNotice>()
export {
  PNotice
};

let noticeId = 0

// Declare install function executed by Vue.use()
export default {
  install(vue: typeof Vue, options?: any) {
    if ((this as any).installed) {
      return;
    }
    (this as any).installed = true;

    vue.prototype.$notify = (options: any) => {
      switch (options.position) {
        case 'left-top':
        case 'top-left':
          handleNotify('left-top', options)
          break;
        case 'right-top':
        case 'top-right':
          handleNotify('right-top', options)
          break;
        case 'left-bottom':
        case 'bottom-left':
          handleNotify('left-top', options)
          break;
        case 'right-bottom':
        case 'bottom-right':
          handleNotify('right-bottom', options)
          break;
        case 'top':
          handleNotify('top', options)
          break;
        case 'bottom':
          handleNotify('bottom', options)
          break;
        default:
          handleNotify('left-top', options)
      }
    }
  },
};

function handleNotify(position: string, options: any) {

  let notice = cache.get(position)
  if (notice) {
    notice.$props.noticeList.push(_getNotice(options))
  } else {
    const instance = new Notice({ propsData: { position, noticeList: [_getNotice(options)] } })
    instance.$mount()
    cache.set(position, instance)
    document.body.appendChild(instance.$el)
  }
}

function _getNotice(options: any) {
  noticeId++
  delete options.positon
  return { id: noticeId, ...{ duration: 5, dismissble: false, merge: false, show: true }, ...options }
}
