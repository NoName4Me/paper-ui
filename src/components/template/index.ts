import PTemplate from './PTemplate.vue';
import Vue from 'vue';


export {
  PTemplate,
};

// Declare install function executed by Vue.use()
export default {
  install(vue: typeof Vue, options?: any) {
    if ((this as any).installed) {
      return;
    }
    (this as any).installed = true;
    Vue.component('PTemplate', PTemplate);
  },
};
