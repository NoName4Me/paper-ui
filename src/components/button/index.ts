import PButton from './PButton.vue';
import Vue from 'vue';


const uiList = [
  PButton,
];

export {
  PButton,
};

// Declare install function executed by Vue.use()
export default {
  install(vue: typeof Vue, options?: any) {
    if ((this as any).installed) {
      return;
    }
    (this as any).installed = true;
    vue.component('PButton', PButton);
  },
};
