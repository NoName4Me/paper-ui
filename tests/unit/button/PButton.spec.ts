import { shallowMount } from '@vue/test-utils';
import { PButton } from '@/components/button';

const factory = (datas = {}, props = {}) => {
  return shallowMount(PButton, {
    data() {
      return { ...datas };
    },
    propsData: { ...props },
  });
};

describe('PButton.vue', () => {
  // it('renders props.msg when passed', () => {
  //   const msg = 'new message';
  //   const wrapper = factory(undefined, { type: 'text' });
  //   expect(wrapper.text()).toMatch(msg);
  // });

  it('primary button', () => {
    const wrapper = factory(undefined, { type: 'primary' });
    expect(wrapper).toMatchSnapshot();
  });

  it('small button', () => {
    const wrapper = factory(undefined, { size: 'small' });
    expect(wrapper).toMatchSnapshot();
  });

  it('disabled big button', () => {
    const wrapper = factory(undefined, { size: 'large', disabled: true });
    expect(wrapper).toMatchSnapshot();
  });
});
