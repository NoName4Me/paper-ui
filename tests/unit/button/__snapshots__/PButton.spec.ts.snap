// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PButton.vue disabled big button 1`] = `
<button disabled="disabled" class="p-btn large">
  <!----> </button>
`;

exports[`PButton.vue primary button 1`] = `
<button class="p-btn middle primary">
  <!----> </button>
`;

exports[`PButton.vue small button 1`] = `
<button class="p-btn small">
  <!----> </button>
`;
